src/libraries/System.Numerics.Vectors/tests/GenericVectorTests.cs

```diff
@@ -846,6 +846,54 @@ public void VectorSingleEqualsNaNTest()
             var nan = new Vector<float>(float.NaN);
             Assert.True(nan.Equals(nan));
         }
+
+        [Fact]
+        public void VectorDoubleEqualsNonCanonicalNaNTest()
+        {
+            var maxSignificand = (1UL << 53) - 1;
+            // NaN with mantissa bits set
+            var snan = BitConverter.UInt64BitsToDouble(BitConverter.DoubleToUInt64Bits(double.NaN) | maxSignificand);
+            var nans = new double[]
+            {
+                double.CopySign(double.NaN, -0.0), // -qnan
+                double.CopySign(double.NaN, +0.0), // +qnan
+                double.CopySign(snan, -0.0),       // -snan
+                double.CopySign(snan, +0.0),       // +snan
+            };
+
+            // all Vector<double> NaNs .Equals compare the same
+            foreach(var i in nans)
+            {
+                foreach(var j in nans)
+                {
+                    Assert.True(new Vector<double>(i).Equals(new Vector<double>(j)));
+                }
+            }
+        }
+
+        [Fact]
+        public void VectorSingleEqualsNonCanonicalNaNTest()
+        {
+            var maxSignificand = (1U << 24) - 1;
+            // NaN with mantissa bits set
+            var snan = BitConverter.UInt32BitsToSingle(BitConverter.SingleToUInt32Bits(float.NaN) | maxSignificand);
+            var nans = new float[]
+            {
+                float.CopySign(float.NaN, -0.0f), // -qnan
+                float.CopySign(float.NaN, +0.0f), // +qnan
+                float.CopySign(snan, -0.0f),      // -snan
+                float.CopySign(snan, +0.0f),      // +snan
+            };
+
+            // all Vector<float> NaNs .Equals compare the same
+            foreach(var i in nans)
+            {
+                foreach(var j in nans)
+                {
+                    Assert.True(new Vector<float>(i).Equals(new Vector<float>(j)));
```

Comment:
```suggestion
                    Assert.True(new Vector<float>(i).Equals(new Vector<float>(j)));
                    Assert.False(new Vector<float>(i) == new Vector<float>(j));
```

Reply:
> Shouldn't Vector64, Vector128 and Vector256 be handled too?

I can add tests for that in a separate PR - it wouldn't be in this test file anyway.